enclave/src/rpc/rpc_call.rs

```diff
@@ -0,0 +1,111 @@
+/*
+    Copyright 2019 Supercomputing Systems AG
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+pub extern crate alloc;
+
+use alloc::{str, string::String};
+
+use core::marker::{Send, Sync};
+use jsonrpc_core::Result as RpcResult;
+use jsonrpc_core::*;
+
+use substratee_node_primitives::Request;
+use substratee_worker_primitives::DirectRequestStatus;
+
+use crate::rpc::rpc_call_encoder::RpcCallEncoder;
+use crate::rpc::rpc_info::RpcInfo;
+
+pub type RpcMethodImpl = dyn Fn(Request) -> RpcResult<(RpcInfo, bool, DirectRequestStatus)> + Sync;
+
+/// RPC call structure
+pub struct RpcCall<E, F>
+where
+    E: RpcCallEncoder + Send + Sync + 'static,
+    F: Fn(Request) -> RpcResult<(RpcInfo, bool, DirectRequestStatus)> + Sync + ?Sized + 'static,
+{
+    method_name: &'static str,
+    method_impl: &'static F,
+    call_encoder: E,
+}
+
+impl<E, F> RpcCall<E, F>
+where
+    E: RpcCallEncoder + Send + Sync + 'static,
+    F: Fn(Request) -> RpcResult<(RpcInfo, bool, DirectRequestStatus)> + Sync + ?Sized + 'static,
+{
+    pub fn method_name(&self) -> &'static str {
+        self.method_name
+    }
+
+    // FIXME: this produces a warning, because we're not using the call encoder as field,
+    // but merely as associated function in the implementation. However, if we don't have a field,
+    // the compiler gives an error that type parameter 'E' is not used, even though it clearly is
+    pub fn new(name: &'static str, method: &'static F, encoder: E) -> Self {
+        RpcCall {
+            method_name: name,
+            method_impl: method,
+            call_encoder: encoder,
+        }
+    }
+}
```

Comment:
An abstract struct that combines the encoder (usually JSON) and the RPC method implementation